### PR TITLE
CVE-2022-29155: Update openldap to 2.6.3

### PR DESCRIPTION
**Pull request by suzyw-w**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

Updating openldap to 2.6.3
[CVE-2022-29155](https://github.com/advisories/GHSA-9rm2-9q89-9524)
> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: contrib/openldap-cmake/CMakeLists.txt [hunk 1]
```diff
@@ -160,6 +160,15 @@ set(_ldap_srcs
     "${OPENLDAP_SOURCE_DIR}/libraries/libldap/lbase64.c"
     "${OPENLDAP_SOURCE_DIR}/libraries/libldap/msctrl.c"
     "${OPENLDAP_SOURCE_DIR}/libraries/libldap/psearchctrl.c"
+    "${OPENLDAP_SOURCE_DIR}/libraries/libldap/threads.c"
```
Comment (rschu1ze) [inline contrib/openldap-cmake/CMakeLists.txt:163]: Needs an updated version string in l. 35 of this file.

File: contrib/openldap-cmake/CMakeLists.txt [hunk 2]
```diff
@@ -184,43 +193,4 @@ target_compile_definitions(_ldap
...
-    PRIVATE "${OPENLDAP_SOURCE_DIR}/libraries/libldap_r"
-    PRIVATE "${OPENLDAP_SOURCE_DIR}/libraries/libldap"
-)
-
-target_compile_definitions(_ldap_r
-    PRIVATE LDAP_R_COMPILE
-    PRIVATE LDAP_LIBRARY
-)
-
-add_library(ch_contrib::ldap ALIAS _ldap_r)
```
Comment (rschu1ze) [inline contrib/openldap-cmake/CMakeLists.txt:225]: Library alias "ch_contrib::ldap" was removed. It is used in files "src/CMakeLists.txt" and "src/configure_config.cmake". I am pretty sure that things will break.

File: contrib/openldap-cmake/CMakeLists.txt [hunk 3]
```diff
@@ -184,43 +193,4 @@ target_compile_definitions(_ldap
...
-
-target_include_directories(_ldap_r SYSTEM
-    PUBLIC ${_extra_build_dir}/include
-    PUBLIC "${OPENLDAP_SOURCE_DIR}/include"
-    PRIVATE "${OPENLDAP_SOURCE_DIR}/libraries/libldap_r"
-    PRIVATE "${OPENLDAP_SOURCE_DIR}/libraries/libldap"
-)
-
-target_compile_definitions(_ldap_r
-    PRIVATE LDAP_R_COMPILE
```
Comment (rschu1ze) [inline contrib/openldap-cmake/CMakeLists.txt:221]: Just wondering if the files in l. 188-197 needed compile definition "LDAP_R_COMPILE". If yes, that compile definition would need to appear in l. 193 on the right side.

**Comment by rschu1ze** [comment]
[ClickHouse's "OpenLDAP" fork](https://github.com/ClickHouse/openldap) is in a sad state. ClickHouse is build from the tip of a [PR](https://github.com/ClickHouse/openldap/pull/1) which isn't merged since almost two years and which also cannot be merged due to a merge conflict. This update to 2.6.3 references another [PR](https://github.com/ClickHouse/openldap/pull/3) which is build on top of the unmerged PR.

Before this gets too messy, we should split the first [PR](https://github.com/ClickHouse/openldap/pull/1) into two PRs: one which updates to 2.5 and one which uses BoringSSL. Both shall be merged first. I can take care of that.

Afterwards, we can proceed with the update to 2.6.3. It should be based on the "master" branch in the OpenLDAP fork and not reference another PR.

**Comment by SuzyWangIBMer** [comment]
Sounds good, thanks. Will redo once pr is merged.

**Comment by rschu1ze** [comment]
After I messed with openldap, it is unfortunately in an even worse state than before. I reverted
- Update to 2.5 in OpenLDAP fork: https://github.com/ClickHouse/openldap/pull/4 here: https://github.com/ClickHouse/openldap/pull/7
- Use BoringSSL in OpenLDAP fork: https://github.com/ClickHouse/openldap/pull/5 here: https://github.com/ClickHouse/openldap/pull/6

Then I tried to reset the OpenLdap fork via `git reset --hard` to 2.5.13: https://github.com/ClickHouse/openldap/pull/8 but
1. when I try to build ClickHouse from that, I get linker errors that some OpenSSL symbols cannot be found (despite libldap and liblber linking OpenSSL),
3. https://github.com/ClickHouse/openldap/pull/8 has some unexpected merge conflicts (either a problem with Git messing around or because the sources in the fork repository were modified before they were initially checked in), and
4. 3. I am afraid that the platform-specific files in ClickHouse's "contrib/openldap-cmake/linux_x86_64/", "contrib/openldap-cmake/linux_aarch64/", etc. no longer fit to OpenLDAP 2.5 as they were generated by hand from OpenLDAP 2.4.

I am afraid the 2.5 upgrade needs to be done completely from scratch. Also not sure if the existing fork repository https://github.com/ClickHouse/openldap is the right repository (looking at problem 2. above) 😞

**Comment by SuzyWangIBMer** [comment]
1. In the latest openldap repository, they have moved all files under libldap_r to libldap. I think this is causing the linker error. CMakeList.txt needs to be updated regarding this change. I have made those changes in this pr. 

Regarding 2&3, if you can confirm which repository is the right one and how the team want to deal with platform-specific files, I can help do 2.5 update as well. We need this library updated to the latest to get rid of CVE errors([CVE-2022-29155](https://github.com/advisories/GHSA-9rm2-9q89-9524)).

**Comment by alexey-milovidov** [comment]
This vulnerability of the OpenLDAP library is not exploitable with ClickHouse.

Firstly, the offending code is simply not included in the build.
Secondly, even if it was included, it will require `dlopen` to operate, and we have special protection to disallow any usage of `dlopen`.

**Comment by bkuschel** [comment]
Thanks Alexey, this PR is more in the spirit of keeping dependencies fresh to avoid scanning tools (who have no idea about what code is being used or not) from flagging ClickHouse as being affected by a high severity CVE.  If it's too complicated to upgrade, we can deal with it.

**Pull request closed without merging**